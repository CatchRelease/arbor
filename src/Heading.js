import styled, { css } from 'react-emotion';
import {
  brandFont,
  fontSizes,
  fontWeights,
  lineHeights,
  colors
} from './theme';

const lineHeightMultiplier = 0.5;

const calculatedHeadingStyles = size =>
  css({
    fontSize: `${size}rem`,
    marginBottom: `calc(${size}rem * ${lineHeightMultiplier})`
  });

const Heading = styled.h1`
  font-family: '${brandFont}';
  font-weight: ${fontWeights.bold};
  line-height: ${lineHeights.small};
`;

Heading.h1 = styled(Heading.withComponent('h1'))`
  ${calculatedHeadingStyles(fontSizes.size7)};
  color: ${colors.grey100};
`;

Heading.h2 = styled(Heading.withComponent('h2'))`
  ${calculatedHeadingStyles(fontSizes.size6)};
  color: ${colors.grey80};
`;

Heading.h3 = styled(Heading.withComponent('h3'))`
  ${calculatedHeadingStyles(fontSizes.size5)};
  color: ${colors.grey80};
`;

Heading.h4 = styled(Heading.withComponent('h4'))`
  ${calculatedHeadingStyles(fontSizes.size4)};
  color: ${colors.grey80};
`;

Heading.h5 = styled(Heading.withComponent('h5'))`
  ${calculatedHeadingStyles(fontSizes.size2)};
  color: ${colors.grey100};
  font-weight: ${fontWeights.medium};
  line-height: ${lineHeights.large};
`;

Heading.h6 = styled(Heading.withComponent('h6'))`
  ${calculatedHeadingStyles(fontSizes.size2)};
  color: ${colors.grey80};
  font-weight: ${fontWeights.regular};
  line-height: ${lineHeights.large};
`;

export default Heading;
