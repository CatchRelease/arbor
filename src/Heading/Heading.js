import React from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import Text from '../Text';

const lineHeightMultiplier = 0.5;

const baseStyles = ({ theme }) => css`
  line-height: ${theme.lineHeights.small};
`;

const calculatedMargin = ({ mb, fontSize, theme }) =>
  !mb &&
  css`
    margin-bottom: calc(${theme.fontSizes[fontSize]} * ${lineHeightMultiplier});
  `;

const Heading = styled(Text)`
  ${baseStyles};
  ${calculatedMargin};
`;

Heading.defaultProps = {
  as: 'h1',
  color: 'text.dark',
  fontWeight: 'bold',
  lineheight: 'small'
};

const createHeading = h => (
  { children, ...props } // eslint-disable-line react/prop-types
) => (
  <Heading as={h} {...props}>
    {children}
  </Heading>
);

Heading.h1 = createHeading('h1');
Heading.h1.defaultProps = {
  fontSize: 'size7'
};

Heading.h2 = createHeading('h2');
Heading.h2.defaultProps = {
  fontSize: 'size6'
};

Heading.h3 = createHeading('h3');
Heading.h3.defaultProps = {
  fontSize: 'size5'
};

Heading.h4 = createHeading('h4');
Heading.h4.defaultProps = {
  fontSize: 'size4'
};

Heading.h5 = styled(createHeading('h5'))`
  line-height: ${props => props.theme.lineHeights.large};
`;
Heading.h5.defaultProps = {
  fontSize: 'size4',
  fontWeight: 'medium'
};

Heading.h6 = styled(createHeading('h6'))`
  line-height: ${props => props.theme.lineHeights.large};
`;
Heading.h6.defaultProps = {
  fontSize: 'size4',
  fontWeight: 'regular'
};

export default Heading;
