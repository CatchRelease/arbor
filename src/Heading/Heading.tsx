import { ElementType, FC } from 'react';
import { system } from 'styled-system';
import type { SetOptional } from 'type-fest';
import styled from '@emotion/styled';
import { css, Theme } from '@emotion/react';

import { FontSize, FontWeight } from '../theme/typography';
import Text, { TextProps } from '../Text';

const lineHeightMultiplier = 0.5;

const baseStyles = ({ theme }: { theme: Theme }) => css`
  line-height: ${theme.lineHeights.small};
`;

const calculatedMargin = ({ mb, theme }: { mb?: string; theme: Theme }) =>
  !mb &&
  system({
    fontSize: {
      property: 'marginBottom',
      transform: (fontSize: FontSize) => {
        return `calc(${theme.fontSizes[fontSize]} * ${lineHeightMultiplier})`;
      }
    }
  });

export type Props = TextProps & {
  fontWeight?: FontWeight;
  mb?: string;
};

const Heading = styled(Text)<Props>`
  ${baseStyles};
  ${calculatedMargin};
`;

Heading.defaultProps = {
  as: 'h1',
  color: 'text.dark',
  fontWeight: 'bold',
  lineHeight: 'small'
};

const createHeading =
  (
    as: ElementType,
    defaultFontSize: FontSize
  ): FC<SetOptional<Props, 'fontSize'>> =>
  ({ children, fontSize = defaultFontSize, ...props }) =>
    (
      <Heading as={as} fontSize={fontSize} {...props}>
        {children}
      </Heading>
    );

const H1 = createHeading('h1', 'size7');

const H2 = createHeading('h2', 'size6');

const H3 = createHeading('h3', 'size5');

const H4 = createHeading('h4', 'size4');

const H5 = styled(createHeading('h5', 'size4'))`
  line-height: ${(props) => props.theme.lineHeights.large};
`;
H5.defaultProps = {
  fontWeight: 'medium'
};

const H6 = styled(createHeading('h6', 'size4'))`
  line-height: ${(props) => props.theme.lineHeights.large};
`;
H6.defaultProps = {
  fontWeight: 'regular'
};

export default Object.assign(Heading, { H1, H2, H3, H4, H5, H6 });
