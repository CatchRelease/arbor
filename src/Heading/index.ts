export { default, Props as HeadingProps } from './Heading';
