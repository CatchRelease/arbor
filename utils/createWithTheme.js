import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
