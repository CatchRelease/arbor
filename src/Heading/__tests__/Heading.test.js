import React from 'react';
import createWithTheme from '../../../utils/createWithTheme';

import Heading from '../Heading';

['H1', 'H2', 'H3', 'H4', 'H5', 'H6'].forEach((variant) => {
  test(`renders a ${variant}`, () => {
    const Variant = Heading[variant];
    const tree = createWithTheme(<Variant>Hello World</Variant>);

    expect(tree).toMatchSnapshot();
  });
});
