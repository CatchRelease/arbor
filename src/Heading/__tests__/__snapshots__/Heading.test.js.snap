// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a h1 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1.5rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1.5rem * 0.5);
}

<h1
  className="emotion-0 emotion-1"
  color="text.dark"
  fontSize="size7"
  fontWeight="bold"
>
  Hello World
</h1>
`;

exports[`renders a h2 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1.25rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1.25rem * 0.5);
}

<h2
  className="emotion-0 emotion-1"
  color="text.dark"
  fontSize="size6"
  fontWeight="bold"
>
  Hello World
</h2>
`;

exports[`renders a h3 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 1rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(1rem * 0.5);
}

<h3
  className="emotion-0 emotion-1"
  color="text.dark"
  fontSize="size5"
  fontWeight="bold"
>
  Hello World
</h3>
`;

exports[`renders a h4 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  color: #0F0F10;
  font-size: 0.875rem;
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: calc(0.875rem * 0.5);
}

<h4
  className="emotion-0 emotion-1"
  color="text.dark"
  fontSize="size4"
  fontWeight="bold"
>
  Hello World
</h4>
`;

exports[`renders a h5 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
  font-weight: 500;
  line-height: 1.25;
  margin-bottom: calc(0.875rem * 0.5);
  line-height: 1.5;
}

<h5
  className="emotion-0 emotion-1"
  fontSize="size4"
  fontWeight="medium"
>
  Hello World
</h5>
`;

exports[`renders a h6 1`] = `
.emotion-0 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25;
  margin-bottom: calc(0.875rem * 0.5);
  line-height: 1.5;
}

<h6
  className="emotion-0 emotion-1"
  fontSize="size4"
  fontWeight="regular"
>
  Hello World
</h6>
`;
